### PR TITLE
Add Uni660 Keyboard

### PR DESCRIPTION
**Pull request by danielhklein**
## Description

Adds the Uni660 QMK Code into the keyboards/sirius directory

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [x] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

n/a

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/sirius/uni660/config.h [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#ifndef CONFIG_H
```
Comment (drashna) [inline keyboards/sirius/uni660/config.h]:
```suggestion
#pragma once
```

File: keyboards/sirius/uni660/config.h [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+#define CONFIG_H
```
Comment (drashna) [inline keyboards/sirius/uni660/config.h]:
```suggestion
```

File: keyboards/sirius/uni660/config.h [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+
+#include "config_common.h"
+
+/* USB Device descriptor parameter */
+
+#define VENDOR_ID       0x5352			// "SR"
+#define PRODUCT_ID      0x0201			// Second Product First Version
```
Comment (drashna) [inline keyboards/sirius/uni660/config.h]: ✅  This VID/PID doesn't appear to be assigned yet.  Looks good.

File: keyboards/sirius/uni660/config.h [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+// and avoid loading invalid data from the EEPROM
+#define EEPROM_VERSION 0x08
+#define EEPROM_VERSION_ADDR 34
+
+// Dynamic keymap starts after EEPROM version
+#define DYNAMIC_KEYMAP_EEPROM_ADDR 35
+// Dynamic macro starts after dynamic keymaps (35+(4*16*5*2)) = (35+640)
+#define DYNAMIC_KEYMAP_MACRO_EEPROM_ADDR 675    // **** CHANGE THIS BASED ON MATRIX_ROWS & MATRIX_COLS ****
+#define DYNAMIC_KEYMAP_MACRO_EEPROM_SIZE 349    // **** CHANGE THIS BASED ON 1024-DYNAMIC_KEYMAP_MACRO_EEPROM_ADDR ****
+#define DYNAMIC_KEYMAP_MACRO_COUNT 16
```
Comment (drashna) [inline keyboards/sirius/uni660/config.h]: This needs to be reconfigured. 

This overlaps with existing ranges, and is configured differently now.
https://caniusevia.com/docs/configuring_qmk/

File: keyboards/sirius/uni660/keymaps/default/keymap.c [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+KC_F3,    KC_LSFT, KC_Z, KC_X, KC_C, KC_V, KC_B, KC_N, KC_M, KC_COMM, KC_DOT, KC_SLSH, KC_RSFT, KC_UP,
+KC_F4,    KC_LCTL, KC_LGUI, KC_LALT, KC_SPC, KC_DEL, KC_SPC, KC_RALT, KC_RGUI, KC_RCTL, KC_LEFT, KC_DOWN, KC_RIGHT 
+),
+};
+
+void matrix_scan_user(void) {
+};
+
+void matrix_init_user(void) {
+};
```
Comment (drashna) [inline keyboards/sirius/uni660/keymaps/default/keymap.c]: These aren't being used and can be removed.

File: keyboards/sirius/uni660/matrix.c [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+        //shifting and transferring the keystates to the QMK matrix variable
+        for (uint8_t i = 0; i < MATRIX_ROWS; i++) {
+            matrix[i] = (uint16_t) uart_data[i*2] | (uint16_t) uart_data[i*2+1] << 8;
+        }
+    }
+
+
+    matrix_scan_quantum();
+    return 1;
+}
```
Comment (drashna) [inline keyboards/sirius/uni660/matrix.c]: No debouncing done here?

Comment (danielhklein) [reply keyboards/sirius/uni660/matrix.c]: is that necessary? I thought QMK implemented debouncing by default, or does matrix_scan override that? If so, are there examples of debouncing on other boards' code I could use as an example?

File: keyboards/sirius/uni660/config.h (hunk 4 above)
Comment (danielhklein) [reply keyboards/sirius/uni660/config.h]: looking into this

Comment (danielhklein) [reply keyboards/sirius/uni660/config.h]: looks like this is old code that should've been removed, taken care of now.

File: keyboards/sirius/uni660/uni660.h [hunk 7]
```diff
@@ -0,0 +1,63 @@
+#ifndef UNI660_H
```
Comment (fauxpark) [inline keyboards/sirius/uni660/uni660.h]:
```suggestion
#pragma once
```

File: keyboards/sirius/uni660/uni660.h [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+#define UNI660_H
```
Comment (fauxpark) [inline keyboards/sirius/uni660/uni660.h]:
```suggestion
```

File: keyboards/sirius/uni660/uni660.h [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+  k40,    k41,   k42,  k43,    k44,  k45,            k48,       k49,   k4a,  k4c,     k4d, k4e, k4f   \
+) \
+  {                                                           \
+    { k00, k01, k02, k03, k04, k05, k06,  KC_NO, k08, k09, k0a, k0b, k0c, k0d, k0e, k0f   }, \
+    { k10, k11, k12, k13, k14, k15, k16,  KC_NO, k18, k19, k1a, k1b, k1c, k1d, k1e, k1f   }, \
+    { k20, k21, k22, k23, k24, k25, k26,  KC_NO, k28, k29, k2a, k2b, k2c, k2d, k2e, k2f   }, \
+    { k30, k31, k32, k33, k34, k35, k36,  KC_NO, k38, k39, k3a, k3b, k3c, k3d, k3e, k3f   }, \
+    { k40, k41, k42, k43, k44, k45, k46,  KC_NO, k48, k49, k4a, KC_NO, k4c, k4d, k4e, k4f    }  \
+  }
+#endif
```
Comment (fauxpark) [inline keyboards/sirius/uni660/uni660.h]:
```suggestion
```

File: keyboards/sirius/uni660/rules.mk [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+F_USB = $(F_CPU)
+
+# Bootloader
+#     This definition is optional, and if your keyboard supports multiple bootloaders of
+#     different sizes, comment this out, and the correct address will be loaded 
+#     automatically (+60). See bootloader.mk for all options.
+BOOTLOADER = caterina
+
+# Interrupt driven control endpoint task(+60)
+OPT_DEFS += -DINTERRUPT_CONTROL_ENDPOINT
```
Comment (fauxpark) [inline keyboards/sirius/uni660/rules.mk]: Could you replace all of this with simply:
```make
# Bootloader selection
#   Teensy       halfkay
#   Pro Micro    caterina
#   Atmel DFU    atmel-dfu
#   LUFA DFU     lufa-dfu
#   QMK DFU      qmk-dfu
#   ATmega32A    bootloadHID
#   ATmega328P   USBasp
BOOTLOADER = caterina
```

File: keyboards/sirius/uni660/rules.mk [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+
+# Build Options
+#   comment out to disable the options.
+#
+BOOTMAGIC_ENABLE = lite	# Virtual DIP switch configuration(+1000)
```
Comment (fauxpark) [inline keyboards/sirius/uni660/rules.mk]:
```suggestion
BOOTMAGIC_ENABLE = lite	# Virtual DIP switch configuration
```

File: keyboards/sirius/uni660/rules.mk [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+MOUSEKEY_ENABLE = no	# Mouse keys(+4700)
```
Comment (fauxpark) [inline keyboards/sirius/uni660/rules.mk]:
```suggestion
MOUSEKEY_ENABLE = no	# Mouse keys
```

File: keyboards/sirius/uni660/rules.mk [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+EXTRAKEY_ENABLE = yes	# Audio control and System control(+450)
```
Comment (fauxpark) [inline keyboards/sirius/uni660/rules.mk]:
```suggestion
EXTRAKEY_ENABLE = yes	# Audio control and System control
```

File: keyboards/sirius/uni660/rules.mk [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+CONSOLE_ENABLE = yes	# Console for debug(+400)
```
Comment (fauxpark) [inline keyboards/sirius/uni660/rules.mk]:
```suggestion
CONSOLE_ENABLE = yes	# Console for debug
```

File: keyboards/sirius/uni660/rules.mk [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+# BACKLIGHT_ENABLE = yes  # Enable keyboard backlight functionality
+# MIDI_ENABLE = yes 		# MIDI controls
+UNICODE_ENABLE = yes 		# Unicode
+# BLUETOOTH_ENABLE = yes # Enable Bluetooth with the Adafruit EZ-Key HID
+
+USB = /dev/ttyACM0
+
+OPT_DEFS += -DMITOSIS_PROMICRO
+MITOSIS_UPLOAD_COMMAND = while [ ! -r $(USB) ]; do sleep 1; done; \
+                         avrdude -p $(MCU) -c avr109 -U flash:w:$(TARGET).hex -P $(USB)
```
Comment (fauxpark) [inline keyboards/sirius/uni660/rules.mk]: This stuff can be deleted - the standard way to flash now is by setting `BOOTLOADER` above, and using the `:flash` make target.

File: keyboards/sirius/uni660/rules.mk [hunk 16]
```diff
@@ -0,0 +1,69 @@
...
+
+# # project specific files
```
Comment (fauxpark) [inline keyboards/sirius/uni660/rules.mk]:
```suggestion
# project specific files
```

File: keyboards/sirius/uni660/keymaps/via/rules.mk [hunk 17]
```diff
@@ -0,0 +1,79 @@
...
+MITOSIS_UPLOAD_COMMAND = while [ ! -r $(USB) ]; do sleep 1; done; \
+                         avrdude -p $(MCU) -c avr109 -U flash:w:$(TARGET).hex -P $(USB)
+
+# # project specific files
+SRC = matrix.c
+
+SRC += keyboards/wilba_tech/wt_main.c
+
+RAW_ENABLE = yes
+DYNAMIC_KEYMAP_ENABLE = yes
```
Comment (fauxpark) [inline keyboards/sirius/uni660/keymaps/via/rules.mk]: None of this is needed anymore.

File: keyboards/sirius/uni660/keymaps/via/keymap.c [hunk 18]
```diff
@@ -0,0 +1,11 @@
...
+
+const uint16_t PROGMEM keymaps[][MATRIX_ROWS][MATRIX_COLS] = {
+[0] = LAYOUT( /* Base */
+KC_ESC,   KC_GRV, KC_1, KC_2, KC_3, KC_4, KC_5, KC_6, KC_7, KC_8, KC_9, KC_0, KC_MINS, KC_EQL, KC_BSPC, KC_DEL, KC_INS,  
+KC_F1,    KC_TAB,  KC_Q, KC_W, KC_E, KC_R, KC_T, KC_Y, KC_U, KC_I, KC_O, KC_P, KC_LBRC, KC_RBRC, KC_BSLS, KC_DEL, 
+KC_F2,    KC_CAPS, KC_A, KC_S, KC_D, KC_F, KC_G, KC_H, KC_J, KC_K, KC_L, KC_SCLN, KC_QUOT, KC_ENT,
+KC_F3,    KC_LSFT, KC_Z, KC_X, KC_C, KC_V, KC_B, KC_N, KC_M, KC_COMM, KC_DOT, KC_SLSH, KC_RSFT, KC_UP,
+KC_F4,    KC_LCTL, KC_LGUI, KC_LALT, KC_SPC, KC_DEL, KC_SPC, KC_RALT, KC_RGUI, KC_RCTL, KC_LEFT, KC_DOWN, KC_RIGHT 
+),
+};
```
Comment (fauxpark) [inline keyboards/sirius/uni660/keymaps/via/keymap.c]: I believe you'll need to add three more layers consisting of `KC_TRNS` to the VIA keymap.

File: keyboards/sirius/uni660/matrix.c (hunk 6 above)
Comment (fauxpark) [reply keyboards/sirius/uni660/matrix.c]: You might be able to use the Custom Matrix Lite feature to get debouncing, and cut down on the amount of boilerplate here: https://docs.qmk.fm/#/custom_matrix?id=_39lite39

File: keyboards/sirius/uni660/keymaps/via/rules.mk [hunk 19]
```diff
@@ -0,0 +1,40 @@
...
+UNICODE_ENABLE = yes 		# Unicode
+# BLUETOOTH_ENABLE = yes # Enable Bluetooth with the Adafruit EZ-Key HID
+
+USB = /dev/ttyACM0
+
+# # project specific files
+SRC = matrix.c
+
+RAW_ENABLE = yes
+DYNAMIC_KEYMAP_ENABLE = yes
```
Comment (fauxpark) [inline keyboards/sirius/uni660/keymaps/via/rules.mk]: All of this can go.

File: keyboards/sirius/uni660/rules.mk [hunk 20]
```diff
@@ -0,0 +1,39 @@
+# Bootloader selection
```
Comment (fauxpark) [inline keyboards/sirius/uni660/rules.mk]:
```suggestion
# MCU name
MCU = atmega32u4

# Bootloader selection
```

File: keyboards/sirius/uni660/rules.mk [hunk 21]
```diff
@@ -0,0 +1,39 @@
...
+#   Teensy       halfkay
+#   Pro Micro    caterina
+#   Atmel DFU    atmel-dfu
+#   LUFA DFU     lufa-dfu
+#   QMK DFU      qmk-dfu
+#   ATmega32A    bootloadHID
+#   ATmega328P   USBasp
+MCU = atmega32u4
```
Comment (fauxpark) [inline keyboards/sirius/uni660/rules.mk]:
```suggestion
```

File: keyboards/sirius/uni660/rules.mk [hunk 22]
```diff
@@ -0,0 +1,39 @@
...
+ARCH = AVR8
```
Comment (fauxpark) [inline keyboards/sirius/uni660/rules.mk]:
```suggestion
```

File: keyboards/sirius/uni660/rules.mk [hunk 23]
```diff
@@ -0,0 +1,39 @@
...
+BOOTLOADER = caterina
+
+# Interrupt driven control endpoint task(+60)
```
Comment (fauxpark) [inline keyboards/sirius/uni660/rules.mk]:
```suggestion
```

File: keyboards/sirius/uni660/rules.mk [hunk 24]
```diff
@@ -0,0 +1,39 @@
...
+OPT_DEFS += -DINTERRUPT_CONTROL_ENDPOINT
```
Comment (fauxpark) [inline keyboards/sirius/uni660/rules.mk]:
```suggestion
```

File: keyboards/sirius/uni660/rules.mk [hunk 25]
```diff
@@ -0,0 +1,39 @@
...
+
```
Comment (fauxpark) [inline keyboards/sirius/uni660/rules.mk]:
```suggestion
```

File: keyboards/sirius/uni660/rules.mk [hunk 26]
```diff
@@ -0,0 +1,39 @@
...
+CUSTOM_MATRIX = yes    # Remote matrix from the wireless bridge
+# Do not enable SLEEP_LED_ENABLE. it uses the same timer as BACKLIGHT_ENABLE
+# SLEEP_LED_ENABLE = yes  # Breathing sleep LED during USB suspend
+NKRO_ENABLE = yes		# USB Nkey Rollover - not yet supported in LUFA
+# BACKLIGHT_ENABLE = yes  # Enable keyboard backlight functionality
+# MIDI_ENABLE = yes 		# MIDI controls
+UNICODE_ENABLE = yes 		# Unicode
+# BLUETOOTH_ENABLE = yes # Enable Bluetooth with the Adafruit EZ-Key HID
+
+USB = /dev/ttyACM0
```
Comment (fauxpark) [inline keyboards/sirius/uni660/rules.mk]:
```suggestion
```

File: keyboards/sirius/uni660/rules.mk [hunk 27]
```diff
@@ -0,0 +1,39 @@
...
+
```
Comment (fauxpark) [inline keyboards/sirius/uni660/rules.mk]:
```suggestion
```

File: keyboards/sirius/uni660/rules.mk [hunk 28]
```diff
@@ -0,0 +1,39 @@
...
+# # project specific files
+SRC = matrix.c
```
Comment (fauxpark) [inline keyboards/sirius/uni660/rules.mk]:
```suggestion
SRC += matrix.c
```

File: keyboards/sirius/uni660/rules.mk [hunk 29]
```diff
@@ -0,0 +1,39 @@
...
+
+RAW_ENABLE = yes
```
Comment (fauxpark) [inline keyboards/sirius/uni660/rules.mk]:
```suggestion
```

File: keyboards/sirius/uni660/rules.mk [hunk 30]
```diff
@@ -0,0 +1,39 @@
...
+DYNAMIC_KEYMAP_ENABLE = yes
```
Comment (fauxpark) [inline keyboards/sirius/uni660/rules.mk]:
```suggestion
```

File: keyboards/sirius/uni660/matrix.c (hunk 6 above)
Comment (drashna) [reply keyboards/sirius/uni660/matrix.c]: Debouncing is handled by the matrix.c file.  If it's not handled here, it doesn't happen. 

Matrix Lite may be a way to go. 

But specifically:
https://github.com/qmk/qmk_firmware/blob/66fe3001e4355d69f5fb8f08ceab9dab793777ef/quantum/matrix.c#L186
And
https://github.com/qmk/qmk_firmware/blob/66fe3001e4355d69f5fb8f08ceab9dab793777ef/quantum/matrix.c#L206

File: keyboards/sirius/uni660/uni660.c [hunk 31]
```diff
@@ -0,0 +1,33 @@
+#include "uni660.h"
+
+void uart_init(void) {
+	SERIAL_UART_INIT();
+}
+
+void led_init(void) {
+	DDRD  |= (1<<1); // Pin to green, set as output
```
Comment (fauxpark) [inline keyboards/sirius/uni660/uni660.c]:
```suggestion
	setPinOutput(D1); // Pin to green, set as output
```

File: keyboards/sirius/uni660/uni660.c [hunk 32]
```diff
@@ -0,0 +1,33 @@
...
+	PORTD |= (1<<1); // Turn it off
```
Comment (fauxpark) [inline keyboards/sirius/uni660/uni660.c]:
```suggestion
	writePinHigh(D1); // Turn it off
```

File: keyboards/sirius/uni660/uni660.c [hunk 33]
```diff
@@ -0,0 +1,33 @@
...
+	DDRF  |= (1<<4) | (1<<5); // Pins to red and blue, set as output
```
Comment (fauxpark) [inline keyboards/sirius/uni660/uni660.c]:
```suggestion
	setPinOutput(F4); // Pins to red and blue, set as output
	setPinOutput(F5);
```

File: keyboards/sirius/uni660/uni660.c [hunk 34]
```diff
@@ -0,0 +1,33 @@
...
+	PORTF |= (1<<4) | (1<<5); // Turn them off
```
Comment (fauxpark) [inline keyboards/sirius/uni660/uni660.c]:
```suggestion
	writePinHigh(F4); // Turn them off
	writePinHigh(F5);
```

File: keyboards/sirius/uni660/uni660.h [hunk 35]
```diff
@@ -0,0 +1,61 @@
+#pragma once
+
+#include "quantum.h"
+
+#define red_led_off   PORTF |= (1<<5)
```
Comment (fauxpark) [inline keyboards/sirius/uni660/uni660.h]:
```suggestion
#define red_led_off   writePinHigh(F5)
```

File: keyboards/sirius/uni660/uni660.h [hunk 36]
```diff
@@ -0,0 +1,61 @@
...
+#define red_led_on    PORTF &= ~(1<<5)
```
Comment (fauxpark) [inline keyboards/sirius/uni660/uni660.h]:
```suggestion
#define red_led_on    writePinLow(F5)
```

File: keyboards/sirius/uni660/uni660.h [hunk 37]
```diff
@@ -0,0 +1,61 @@
...
+#define blu_led_off   PORTF |= (1<<4)
```
Comment (fauxpark) [inline keyboards/sirius/uni660/uni660.h]:
```suggestion
#define blu_led_off   writePinHigh(F4)
```

File: keyboards/sirius/uni660/uni660.h [hunk 38]
```diff
@@ -0,0 +1,61 @@
...
+#define blu_led_on    PORTF &= ~(1<<4)
```
Comment (fauxpark) [inline keyboards/sirius/uni660/uni660.h]:
```suggestion
#define blu_led_on    writePinLow(F4)
```

File: keyboards/sirius/uni660/uni660.h [hunk 39]
```diff
@@ -0,0 +1,61 @@
...
+#define grn_led_off   PORTD |= (1<<1)
```
Comment (fauxpark) [inline keyboards/sirius/uni660/uni660.h]:
```suggestion
#define grn_led_off   writePinHigh(D1)
```

File: keyboards/sirius/uni660/uni660.h [hunk 40]
```diff
@@ -0,0 +1,61 @@
...
+#define grn_led_on    PORTD &= ~(1<<1)
```
Comment (fauxpark) [inline keyboards/sirius/uni660/uni660.h]:
```suggestion
#define grn_led_on    writePinLow(D1)
```

**Comment by danielhklein** [comment]
Are there any other changes that need to be made before this can be merged?

**Comment by drashna** [comment]
Not that I see. It's just awaiting more approvals.

**Comment by fauxpark** [comment]
Thanks!

**Pull request merged**